core/server/worker/src/main/java/alluxio/worker/block/UnderFileSystemBlockReader.java

```diff
@@ -0,0 +1,322 @@
+  private final boolean mNoCache;
+  /** The Alluxio block store. It is used to interact with Alluxio. */
+  private final BlockStore mAlluxioBlockStore;
+
+  /** The input stream to read from UFS. */
+  private InputStream mUnderFileSystemInputStream;
+  /** The block writer to write the block to Alluxio. */
+  private BlockWriter mBlockWriter;
+  /** If set, the reader is closed and should not be used afterwards. */
+  private boolean mClosed;
+
+  /** The position of mUnderFileSystemInputStream (if not null) is blockStart + mPos. */
+  private long mInStreamPos;
+
+  /**
+   * Creates an instance of {@link UnderFileSystemBlockReader} and initializes it with a reading
+   * offset.
+   *
+   * @param blockMeta the block meta
+   * @param offset the position within the block to start the read
+   * @param noCache do not cache the block if set
+   * @param alluxioBlockStore the Alluxio block store
+   * @return the block reader
+   * @throws BlockDoesNotExistException if the UFS block does not exist in the UFS block store
+   * @throws IOException if an I/O related error occur
+   */
+  public static UnderFileSystemBlockReader create(UnderFileSystemBlockMeta blockMeta, long offset,
+      boolean noCache, BlockStore alluxioBlockStore)
+      throws BlockDoesNotExistException, IOException {
+    UnderFileSystemBlockReader ufsBlockReader =
+        new UnderFileSystemBlockReader(blockMeta, noCache, alluxioBlockStore);
+    ufsBlockReader.init(offset);
+    return ufsBlockReader;
+  }
+
+  /**
+   * Creates an instance of {@link UnderFileSystemBlockReader}.
+   *
+   * @param blockMeta the block meta
+   * @param noCache do not cache the block
+   * @param alluxioBlockStore the Alluxio block store
+   */
+  private UnderFileSystemBlockReader(UnderFileSystemBlockMeta blockMeta, boolean noCache,
+      BlockStore alluxioBlockStore) {
+    mFileBufferSize = Configuration.getBytes(PropertyKey.WORKER_FILE_BUFFER_SIZE);
+    mBlockMeta = blockMeta;
+    mAlluxioBlockStore = alluxioBlockStore;
+    mNoCache = noCache;
+    mInStreamPos = -1;
+    mBlockMeta.setBlockReader(this);
+  }
+
+  /**
+   * Initializes the reader. This is only called in the factory method.
+   *
+   * @param offset the position within the block to start the read
+   * @throws BlockDoesNotExistException if the UFS block does not exist in the UFS block store
+   * @throws IOException if an I/O related error occur
+   */
+  private void init(long offset) throws BlockDoesNotExistException, IOException {
+    UnderFileSystem ufs = UnderFileSystem.Factory.get(mBlockMeta.getUnderFileSystemPath());
+    ufs.connectFromWorker(
+        NetworkAddressUtils.getConnectHost(NetworkAddressUtils.ServiceType.WORKER_RPC));
+    if (!ufs.isFile(mBlockMeta.getUnderFileSystemPath())) {
+      throw new BlockDoesNotExistException(
+          ExceptionMessage.UFS_PATH_DOES_NOT_EXIST.getMessage(mBlockMeta.getUnderFileSystemPath()));
+    }
+
+    updateUnderFileSystemInputStream(offset);
```

Comment:
we call 
```java
updateUnderFileSystemInputStream(offset);
updateBlockWriter(offset);
```
together, 
sometimes in the reversed order. Will this matter? Should we be consistent in case error happens we could reason the state?